### PR TITLE
Legend tags are not accepting the full range of display styling possibilities

### PR DESCRIPTION
**Pull request by Ahmad-S792**
#### 3ae9f4142aed981acef64bc585d7525c7dfb74cb
<pre>
Legend tags are not accepting the full range of display styling possibilities

Legend tags are not accepting the full range of display styling possibilities
<a href="https://bugs.webkit.org/show_bug.cgi?id=116031">https://bugs.webkit.org/show_bug.cgi?id=116031</a>

Reviewed by Tim Nguyen.

This patch is to align with Blink / Chrome and Gecko / Firefox.

It changes &apos;legend&apos; element to accept values beyond &quot;display: block&quot; and accept other equivalent of &apos;block&apos; type values as well.

Additionally, it is to also align &quot;User Agent&quot; stylesheet of &quot;legend&quot; tag
by removing &quot;display: block&quot;, which is not in the Web Specification.

This change progresses Webkit on multiple WPT test cases.

* Source/WebCore/style/StyleAdjuster.cpp:
(Adjuster::adjust): Change from &quot;DisplayType::Block&quot; to &quot;equivalentBlockDisplay&quot; for &quot;legend&quot; element
* Source/WebCore/css/html.css: Update &quot;legend&quot; tag to align with web-spec
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/the-fieldset-and-legend-elements/legend-display-expected.txt: Rebaselined
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/the-fieldset-and-legend-elements/legend-grid-flex-multicol-expected.txt: Ditto
* LayoutTests/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/the-fieldset-and-legend-elements/legend-sans-fieldset-display-expected.txt: Ditto
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/the-fieldset-and-legend-elements/legend-sans-fieldset-display-expected.txt: Ditto
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/the-fieldset-and-legend-elements/legend-grid-flex-multicol-expected.txt: Ditto
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/the-fieldset-and-legend-elements/legend-sans-fieldset-display-expected.txt: Added Platform Specific Expectations
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/html/rendering/non-replaced-elements/the-fieldset-and-legend-elements/legend-sans-fieldset-display-expected.txt: Added Platform Specific Expectations

Canonical link: <a href="https://commits.webkit.org/256841@main">https://commits.webkit.org/256841@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/fc485ac515f4b78f4d117c164d84e125526ddf7e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96770 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6034 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29862 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106298 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166588 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100750 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6247 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34767 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89155 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103002 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102445 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4696 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83382 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31638 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86516 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88385 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74573 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/71 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/19871 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/64 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21289 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4735 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4852 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/43818 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1290 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40574 "Passed tests") | |

### CODE REVIEW

File: Source/WebCore/style/StyleAdjuster.cpp [hunk 1]
```diff
@@ -301,7 +301,7 @@ void Adjuster::adjust(RenderStyle& style, const RenderStyle* userAgentAppearance
             }
 
             if (m_element->hasTagName(legendTag))
-                style.setEffectiveDisplay(DisplayType::Block);
+                style.setEffectiveDisplay(equivalentBlockDisplay(style));
```
Comment (nt1m) [inline Source/WebCore/style/StyleAdjuster.cpp:304]: I think this puts us in line with Chrome, though the test says we shouldn't adjust style at all, not sure why we do it in the first place. I think @anttijk has more context on why we do it.

**Comment by Ahmad-S792** [comment]
@nt1m , @smfr - Should we take this to also remove 'display: block' from html.css for 'legend' element since it is not in web-spec. It will align us with web-spec.

File: Source/WebCore/style/StyleAdjuster.cpp (hunk 1 above)
Comment (Ahmad-S792) [reply Source/WebCore/style/StyleAdjuster.cpp:304]: Cool. We also have "display: block" in html.css similar to Chrome for UA stylesheet.

**Comment by Ahmad-S792** [comment]
@anttijk - I have three approaches in mind:

1) Go ahead with current patch and fix tests
2) Go ahead with same patch and also remove "display: block" from html.css
3) Remove whole "legend" from StyleAdjuster and also remove "display: block" from html.css

Let me know  your thoughts and I am happy to amend this accordingly. Since it is helping us progress in WPT. Hence, I am quite interesting it getting fixed. Thanks!

UPDATE - I tried (3) but the number of test failures were around 70, so I opted for (2) and fixed all tests as well.

**Comment by Ahmad-S792** [comment]
@nt1m & @anttijk - Appreciate if you can review this? If it is not right approach, please guide. Since it was progressing Webkit on few WPTs, hence, I opted for this.

**Review by nt1m — APPROVED** [review]
This is more correct because: https://html.spec.whatwg.org/multipage/rendering.html#the-fieldset-and-legend-elements:the-legend-element

says:

> A [fieldset](https://html.spec.whatwg.org/multipage/form-elements.html#the-fieldset-element) element's [rendered legend](https://html.spec.whatwg.org/multipage/rendering.html#rendered-legend), if any, is expected to act as follows:
> * The ['display'](https://drafts.csswg.org/css2/#display-prop) property is expected to behave as if its computed value was blockified.

Although there's quite a bit of work to make it match the spec like Firefox does, e.g.:
* the computed value of `display` shouldn't change (this is what the rest of failures are)
* this should only apply to "rendered legends" (e.g. the legend has to be part of a fieldset)

Thanks!

**Comment by nt1m** [comment]
Also this is the commit that initially added the blockification: https://trac.webkit.org/changeset/41538/webkit

**Pull request merged**